File: src/my_py_pkg/my_py_pkg/my_first_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

class MyNode(Node):

    def __init__(self):
        super().__init__("py_test")
        self.counter_=0
        self.get_logger().info("Hello OOP World! ROS2")
        self.create_timer(0.5, self.timer_callback)    #call callback_fn every 0.5 secs

    def timer_callback(self):
        self.counter_+=1
        self.get_logger().info("Timer Hello " + str(self.counter_))

def main(args=None):
    #first line to write in each ROS2 program. Initialize ROS2 communications
    rclpy.init(args=args)        

    #create node 
    node=MyNode()

    #Pause the program and allow the node to be alive
    rclpy.spin(node)                  #callbacks etc are called from the spin function.
                                      #Publishers/Subscribers go here.


    #last line of program
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"

class MyNode: public rclcpp::Node
{
    public:
        MyNode():Node("cpp_test"),counter_(0)
        {
            //print some info
            RCLCPP_INFO(this->get_logger(),"Hello OOP Cpp World!");

            timer_=this->create_wall_timer(std::chrono::seconds(1),                     //time interval
                                           std::bind(&MyNode::timerCallback, this));    //std::bind needed for calling callback
        }

    private:

        void timerCallback()
        {
            counter_++;
            RCLCPP_INFO(this->get_logger(),"Timer Hello %d",counter_);
        }
        
        //shared pointer to a timer
        rclcpp::TimerBase::SharedPtr timer_;
        int counter_;
};

int main(int argc, char** argv)
{
    //first line to write in each ROS2 program. Initialize ROS2 communications
    rclcpp::init(argc,argv);

    //create node
    auto node=std::make_shared<MyNode>();                  //auto is shared pointer. make_shared returns shared_pointer
                                                           //no need to new and delete
    
    
    //spin the node. keep it alive
    rclcpp::spin(node);

    //last part of the program
    rclcpp::shutdown();

    return 0;


}
